### ISSUE TITLE
compact: use thanos engine or matrix selector for downampling

### ISSUE THREAD

**Issue by rabenhorst** [thanos-io/thanos#6554]
### Is your proposal related to a problem?

The introduction of native histogram support for Thanos has increased complexity as we now manage two metric types instead of one. This complexity is particularly evident in the [unmerged downsampling PR](https://github.com/thanos-io/thanos/pull/6350), which further complicates the already complicated downsampling code. Currently, this code does not support handling metrics with mixed types (i.e., subsequent chunks having different metric types, which for example happens if a histogram and counter have the same name). Incorporating this feature would add another layer of complexity, making the code more challenging to debug and maintain. Additionally, we have duplicated similar functionality in the engine, such as calculating the sum, average, counter, and count, which adds to the overall complexity.

### Describe the solution you'd like

We should rewrite the downsampling code using the Thanos engine or matrix selector for calculation of the aggregates and make the code generic so it can handle both type of metrics.

### Describe alternatives you've considered

Add more code to https://github.com/thanos-io/thanos/pull/6350 so we can handle mixed chunk types or rewrite the code without using the Thanos engine or matrix selector

### Additional context

I will have time to work on poc this week and if it is successful add it to https://github.com/thanos-io/thanos/pull/6350.

**Comment by fpetkovski** [comment]
I agree with this, I feel like the current downsampling approach is a reimplementation of a query engine/recording rules.

It could be much more robust and flexible if it used an actual engine.

**Comment by GiedriusS** [comment]
:+1: sounds like a good idea. I wonder if it would be possible to implement this somehow so that we could switch between those two in the middle of the migration

**Comment by rabenhorst** [comment]
> :+1: sounds like a good idea. I wonder if it would be possible to implement this somehow so that we could switch between those two in the middle of the migration

Should be possible, since input is always a block (reader) from what I understand. I worked on a (very hacky) poc last week, but need to improve tests and will have to look into making querying from aggregates more efficient. After some cleanup and basing it on main, I can link a draft PR here.

**Comment by yeya24** [comment]
This sounds like a very interesting approach.
> We should refactor the downsampling code to utilize the Thanos engine for calculating aggregates and making it more generic regarding the two metric types

I am wondering why only Thanos engine here. Does the existing promql engine work here?

**Comment by rabenhorst** [comment]
> I am wondering why only Thanos engine here. Does the existing promql engine work here?

We can try both. Hopefully next week I can continue working on my PoC. For testing the existing promql engine will be the source of truth anyhow.

**Comment by yeya24** [comment]
I feel it is good to refactor the code to implement this logic now before merging the native histogram downsample PR, to make things easier

**Comment by rabenhorst** [comment]
> I feel it is good to refactor the code to implement this logic now before merging the native histogram downsample PR, to make things easier.
> 
> @rabenhorst did we try the promql engine approach? How did it go

I have experimented a bit (https://github.com/rabenhorst/thanos/tree/downsampling-v2), but never continued the work. IIRC @MichaHoffmann was also looking into it.
